### ISSUE TITLE
Making `/sync` and `/keys/query` requests at the same time

### ISSUE THREAD

**Issue by poljar** [matrix-org/matrix-rust-sdk#1386]
## Sync setup and problem description
Our current setup to track devices of users[[1]] assumes that we don't run a `/sync` request in parallel. Our setup has the following steps in a loop:

1. Sync with the server
2. Pass the E2EE related changes to the `OlmMachine` via `receive_sync_changes()`
3. Process the rest of the sync, calling `OlmMachine::update_tracked_users()` when needed
4. Send the outgoing requests from `OlmMachine::outgoing_requests()` to the server

This is a bit simplified, but captures the gist of it.

Step 3 will signal the `OlmMachine` that we need to track the devices of the users that were passed in as the argument to the `update_tracked_users()` method. The `OlmMachine` then proceeds to save a tuple of `(UserId, bool)`, this corresponds to the two flags mentioned in the spec[[1]]:

> It first sets a flag to record that it is now tracking Bob’s device list, and a separate flag to indicate that its list of Bob’s devices is outdated.

Step 4 then will make a `/keys/query` request for all the users we are tracking devices for, and has the boolean `outdated` flag set to `true`. After a successful `/keys/query` response has been processed for a given user, the `outdated` flag will be set to `false`.

The boolean, `outdated`, flag can be set to `true` again by a `/sync` response from step 1.

This means that we can't process a `/keys/query` at the same time we process a `/sync` as we're running into a race where each of them might want to overwrite each others `outdated` boolean.

Sliding sync, and likely all other client setups, don't want to wait for the processing of the `OlmMachine::outgoing_request()` before they make the next `/sync` request. 

Another problem to tackle is that, as per spec[[1]], we don't track the devices of all the users we're sharing a room with. We only track the devices of the users that we're sharing an encrypted room with, yet clients have features that let you inspect the devices of all users.

## Tl;dr and proposed solution
To reiterate, we have two problems:

1. We can't safely let `/sync` happen at the same time a `/keys/query` is going on.
2. We can't safely let clients inspect the devices of a user that we're not tracking the device list for.

Problem 1. could be tackled by replacing the `outdated` boolean with a counter, where `/sync` may increment the counter and a `/keys/query` response would decrement it. The counter could be clamped at some value to prevent fast `/sync` responses from incrementing it needlessly. The clamping isn't strictly needed, but would also clamp the amount of needless `/keys/query` requests we send out.

Problem 2 will be further discussed in #1191, but I think it might boil down either:
1. Don't store devices of users we're not tracking the device list for in the DB
2. Store the devices in a separate bucket, so we know that they might be outdated.
   

[1]: https://spec.matrix.org/v1.5/client-server-api/#tracking-the-device-list-for-a-user
